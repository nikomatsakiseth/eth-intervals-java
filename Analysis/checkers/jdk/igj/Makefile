
.SILENT:

# Variables
CHECKERS ?= ../..
CHECKERS_JAR ?= $(CHECKERS)/checkers.jar

LANGTOOLS ?= ../../../jsr308-langtools
JAVADOCLIB ?= $(LANGTOOLS)/build/classes
JAVAC ?= $(LANGTOOLS)/dist/bin/javac -Xlint
JAVAP ?= $(LANGTOOLS)/dist/bin/javap

RTJAR ?= $(JAVA_HOME)/jre/lib/rt.jar

clean:
	rm -rf build annotated jdk.jar

build:
	mkdir -p build


############## Validate annotated jdk against missing methods ##############
SOURCES = $(shell find src/java -name '*.java')
CLASSES = $(patsubst src/%.java,build/%.class, $(SOURCES))
DIFFS = $(patsubst src/%.java,build/%.diff, $(SOURCES))

.PHONY: classes
classes: $(SOURCES)
	echo "building the jdk..."
	mkdir -p build
	$(JAVAC) -cp $(CHECKERS_JAR):$(JAVADOCLIB) -d build $?

build/%.class: src/%.java
	$(JAVAC) -sourcepath src -cp $(CHECKERS_JAR):$(JAVADOCLIB) -d build $<

build/%.expected: build/%.class
	$(JAVAP) -cp $(JAVADOCLIB) $(subst /,.,$*)  | grep public | sed 's/ native / /g' | sed 's/? extends java\/lang\/Object/?/g' | sort > $@

build/%.actual: build/%.class
	$(JAVAP) -Xbootclasspath/p:build $(subst /,.,$*)  | grep public | sed 's/ native / /g' | sed 's/? extends java\/lang\/Object/?/g' | sort > $@

build/%.diff: build/%.expected build/%.actual
	echo "validating class:" $(subst /,.,$*)
	diff $? > $@

validate: classes $(DIFFS)

############## Annotate the annotated jdk file ##############
JAIFS = $(patsubst src/%.java,build/%.jaif, $(SOURCES))
ANNOTATED = $(patsubst src/%.java,annotated/%.class, $(SOURCES))

build/%.jaif: build/%.class
	extract-annotations $<

stubs: $(JAIFS)

annotated/%.class: build/%.jaif
	cd annotated; unzip $(RTJAR) $*.class > /dev/null
	insert-annotations $@ $<

annotated: $(ANNOTATED)

jdk.jar: annotated
	cd annotated; jar cf ../jdk.jar `find . -name '*.class'`

jdk: jdk.jar
