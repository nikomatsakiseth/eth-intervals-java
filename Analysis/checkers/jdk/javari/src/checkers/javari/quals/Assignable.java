package checkers.javari.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.javari.JavariChecker;

import static java.lang.annotation.ElementType.*;

/**
 * Indicates that a field is assignable, even if it is inside a {@link ReadOnly}
 * instance.
 *
 * <p>
 *
 * This annotation is part of the Javari language.
 *
 * @see JavariChecker
 * @checker.framework.manual #javari-checker Javari Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({FIELD})
public @interface Assignable {

}
