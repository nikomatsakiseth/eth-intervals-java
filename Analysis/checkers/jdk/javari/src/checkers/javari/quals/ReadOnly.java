package checkers.javari.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.javari.JavariChecker;
import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;
import static java.lang.annotation.ElementType.*;

/**
 * Indicates that, for the variable on which this annotation appears,
 * the object to which this variable refers will not be modified via
 * this reference, except its fields explicitly marked as
 * {@link Mutable}.
 *
 * <p>
 *
 * This annotation is part of the Javari language.
 *
 * @see Mutable
 * @see JavariChecker
 * @checker.framework.manual #javari-checker Javari Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
//@Target({FIELD, LOCAL_VARIABLE, METHOD, PARAMETER, TYPE})
@TypeQualifier
@SubtypeOf({})
public @interface ReadOnly {

}
