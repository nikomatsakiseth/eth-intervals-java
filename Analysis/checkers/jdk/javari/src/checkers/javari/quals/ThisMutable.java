package checkers.javari.quals;

import java.lang.annotation.Target;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;

/**
 * An annotation used to represent a place holder immutability type, that is
 * equivalent to the ThisMutable type on the Javari typesystem.
 *
 * However, it is an implementation detail; hence, the package-scope.
 */
@TypeQualifier
@Target({}) // A programmer cannot write @ThisMutable in a program
@SubtypeOf(ReadOnly.class)
public @interface ThisMutable {

}
