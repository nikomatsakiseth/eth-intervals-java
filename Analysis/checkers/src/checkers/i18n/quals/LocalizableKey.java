package checkers.i18n.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;
import checkers.quals.Unqualified;

/**
 * Indicates that the {@code String} type has been localized and
 * formatted for the target output locale
 */
@TypeQualifier
@SubtypeOf(Unqualified.class)
@Documented
@Retention(RetentionPolicy.RUNTIME)
public @interface LocalizableKey { }
