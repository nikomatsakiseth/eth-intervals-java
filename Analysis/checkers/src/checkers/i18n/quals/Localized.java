package checkers.i18n.quals;

import static com.sun.source.tree.Tree.Kind.*;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import checkers.quals.*;

/**
 * Indicates that the {@code String} type has been localized and
 * formatted for the target output locale
 */
@TypeQualifier
@SubtypeOf(Unqualified.class)
@ImplicitFor( trees = {
        /* All integer literals */
        INT_LITERAL,
        LONG_LITERAL,
        FLOAT_LITERAL,
        DOUBLE_LITERAL,
        BOOLEAN_LITERAL,

        //CHAR_LITERAL,
        //STRING_LITERAL,
        //NULL_LITERAL
        }
)
@Documented
@Retention(RetentionPolicy.RUNTIME)
public @interface Localized { }
