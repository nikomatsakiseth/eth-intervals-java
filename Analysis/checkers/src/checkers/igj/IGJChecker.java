package checkers.igj;

import javax.annotation.processing.*;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.Element;

import com.sun.source.tree.ExpressionTree;
import com.sun.source.tree.Tree;

import checkers.basetype.BaseTypeChecker;
import checkers.igj.quals.*;
import checkers.quals.TypeQualifiers;
import checkers.types.*;
import checkers.types.AnnotatedTypeMirror.AnnotatedDeclaredType;
import checkers.util.*;

/**
 * A type-checker plug-in for the IGJ immutability type system that finds (and
 * verifies the absence of) undesired side-effect errors.
 *
 * The IGJ language is a Java language extension that expresses immutability
 * constraints, using six annotations: {@link ReadOnly}, {@link Mutable},
 * {@link Immutable}, {@link I} -- a polymorphic qualifier, {@link Assignable},
 * and {@link AssignsFields}.  The language is specified by the FSE 2007 paper.
 *
 * @checker.framework.manual #igj-checker IGJ Checker
 *
 */
@TypeQualifiers({ ReadOnly.class, Mutable.class, Immutable.class, I.class,
    AssignsFields.class, IGJBottom.class })
public class IGJChecker extends BaseTypeChecker {
    //
    // IGJ tries to adhere to the various rules specified by the
    // type system and the conventions of the framework, except for two
    // things:
    // 1. overloading the meaning of BOTTOM_QUAL
    //    Review the javadoc of #createQualiferHierarchy
    //
    // 2. Having two qualifiers for a given type in one particular case
    //    which is that the self type (i.e. type of 'this' identifier) within
    //    a method with an AssignsFields receiver within I classes, then the self type is
    //    '@AssignsFields @I EnclosingClass' and they are treated as
    //    Incomparable.  This is useful in the following cases:
    //
    //    a. for method invocability tests, a method with an AssignsFields receiver from within
    //       a readonly context can be called only via AssignsFields reference
    //       of 'this'.  I cannot be a receiver type, so it doesn't interfere.
    //
    //    b. for assignment, 'this' can be assigned to '@I EnclosingClass'
    //       reference within such methods (assignment encompasses the escape
    //       of this when passed to method parameters).  Fields and variables
    //       cannot be AssignsFields, so it's safe.
    //
    //    The design of QualifierHierarchy.isSubtype(Collection, Collection)
    //    reflect this choice.
    //
    /** Supported annotations for IGJ.  Used for subtyping rules. **/
    protected AnnotationMirror READONLY, MUTABLE, IMMUTABLE, I, ASSIGNS_FIELDS, BOTTOM_QUAL;

    @Override
    public synchronized void init(ProcessingEnvironment env) {
        AnnotationUtils annoFactory = AnnotationUtils.getInstance(env);
        READONLY = annoFactory.fromClass(ReadOnly.class);
        MUTABLE = annoFactory.fromClass(Mutable.class);
        IMMUTABLE = annoFactory.fromClass(Immutable.class);
        I = annoFactory.fromClass(I.class);
        ASSIGNS_FIELDS = annoFactory.fromClass(AssignsFields.class);
        BOTTOM_QUAL = annoFactory.fromClass(IGJBottom.class);
        super.init(env);
    }

    // **********************************************************************
    // IGJ specific Type Relationship
    // **********************************************************************

    /**
     * Return true if the assignment variable is an assignable field or
     * variable, and returns false otherwise.
     *
     * A field is assignable if it is
     *
     * 1. a static field
     * 2. marked {@link Assignable}
     * 3. accessed through a mutable reference
     * 4. reassigned with an {@link AssignsFields} method and owned by 'this'
     *
     */
    @Override
    public boolean isAssignable(AnnotatedTypeMirror varType,
            AnnotatedTypeMirror receiverType, Tree varTree) {
        if (!(varTree instanceof ExpressionTree))
            return true;

        Element varElement = InternalUtils.symbol(varTree);
        if (varTree.getKind() != Tree.Kind.ARRAY_ACCESS
                && (varElement == null // a variable element should never be null
                        || !varElement.getKind().isField()
                        || ElementUtils.isStatic(varElement)
                        || varElement.getAnnotation(Assignable.class) != null))
            return true;

        assert receiverType != null;

        final boolean isAssignable =
            (receiverType.hasAnnotation(MUTABLE)
             || receiverType.hasAnnotation(BOTTOM_QUAL)
             || (receiverType.hasAnnotation(ASSIGNS_FIELDS)
                     && TreeUtils.isSelfAccess((ExpressionTree)varTree)));

        return isAssignable;
    }

    @Override
    public boolean isValidUse(AnnotatedDeclaredType elemType, AnnotatedDeclaredType use) {
        if (elemType.hasAnnotation(I) || use.hasAnnotation(READONLY))
            return true;
        return super.isValidUse(elemType, use);
    }

    // **********************************************************************
    // Factory methods
    // **********************************************************************

    @Override
    protected QualifierHierarchy createQualifierHierarchy() {
        return new IGJQualifierHierarchy((GraphQualifierHierarchy)super.createQualifierHierarchy());
    }

    @Override
    protected TypeHierarchy createTypeHierarchy() {
        return new IGJTypeHierarchy(getQualifierHierarchy());
    }

    //
    // IGJ makes an interesting use of BOTTOM_QUAL (IGJBottom).  It gets used
    // in two ways:
    //
    // __ AS BOTTOM QUALIFIER __
    //
    // This is the intended use and design for it
    //
    // A bottom qualifier is needed to annotate some expressions, like 'null'.
    // Otherwise, null would need to be '@Mutable @Immutable <nulltype>' so
    // it could be assigned to everything.
    //
    // __ AS SUPER QUALIFIER __
    //
    // As I used IGJ, I realized that it's useful to have an annotation
    // that acts as a place holder qualifier that is a supertype of
    // everything.  The semantics of such qualifier is a bit
    // different from ReadOnly.
    //
    // It's only because the existence of un-annotated code that we need this
    // annotation, and IGJ's promise that un-annotated code should type check.
    //
    // TODO: Explain these cases more
    //
    private final class IGJQualifierHierarchy extends GraphQualifierHierarchy {
        public IGJQualifierHierarchy(GraphQualifierHierarchy hierarchy) {
            super(hierarchy);
        }
        public boolean isSubtype(AnnotationMirror rhs, AnnotationMirror lhs) {
            return (AnnotationUtils.areSame(rhs, BOTTOM_QUAL)
                    || AnnotationUtils.areSame(lhs, BOTTOM_QUAL)
                    || super.isSubtype(rhs, lhs));
        }
    }

    /**
     * Represents the annotated type hierarchy of the IGJ type system.
     *
     * The IGJ type system diverges from the JLS in two ways:
     * 1. Type arguments are always co-variant with respect to
     *    {@link IGJBottom}
     *
     * 2. If the type is a read-only or an immutable type, then type arguments
     *    may change co-variantly in a safe manner
     */
    private final class IGJTypeHierarchy extends TypeHierarchy {
        public IGJTypeHierarchy(QualifierHierarchy qualifierHierarchy) {
            super(qualifierHierarchy);
        }

        /**
         * Returns true if either of the provided types is a
         * {@link IGJBottom}, otherwise uses the JLS specification
         * implemented by the abstract {@link typeHierarchy}.
         *
         */
        // Note: This cannot be expressed with the QualifierHierarchy alone,
        // as TypeHierarchy requires type arguments to be equivalent
        @Override
        protected boolean isSubtypeAsTypeArgument(AnnotatedTypeMirror rhs, AnnotatedTypeMirror lhs) {
            return (lhs.hasAnnotation(BOTTOM_QUAL)
                    || rhs.hasAnnotation(BOTTOM_QUAL)
                    || super.isSubtypeAsTypeArgument(rhs, lhs));
        }


        /**
         * Uses the JLS specification (as implemented in {@link TypeHierarchy},
         * if the variable type, lhs, is mutable; otherwise, allows the type
         * arguments to change while maintaining subtype relationship.
         *
         * This allows for subtyping relationships of the kind:
         * <pre>  @Mutable List&lt;@Mutable Date&gt; &lt;: @ReadOnly List&lt;@ReadOnly Date&gt;<\pre>
         */
        @Override
        protected boolean isSubtypeTypeArguments(AnnotatedDeclaredType rhs, AnnotatedDeclaredType lhs) {
            if (lhs.hasAnnotation(MUTABLE))
                return super.isSubtypeTypeArguments(rhs, lhs);

            if (!lhs.getTypeArguments().isEmpty()
                    && !rhs.getTypeArguments().isEmpty()) {
                assert lhs.getTypeArguments().size() == rhs.getTypeArguments().size();
                for (int i = 0; i < lhs.getTypeArguments().size(); ++i) {
                    if (!isSubtype(rhs.getTypeArguments().get(i), lhs.getTypeArguments().get(i)))
                        return false;
                }
            }
            return true;
        }
    }
}
