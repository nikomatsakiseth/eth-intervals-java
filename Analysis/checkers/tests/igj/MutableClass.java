import checkers.igj.quals.*;

/**
 *
 * This is a test class for the IGJ Checker framework
 *
 * This class tests the basic operations, such as:
 *
 * <ul>
 *  <li> Mutation through references
 *  <li> Subtyping: Assignments, method invocations
 *  <li> Fields Rule (without any recursion)
 *  <li> method invocation rule (partially)
 * </ul>
 *
 * This class does not test for {@code @AssignsFields} methods
 */
@Mutable
public class MutableClass {
    int field = 5;
    @Assignable int assignable = 4;

    // Fields
    @Mutable MutableClass mutableRef = new @Mutable MutableClass();
    @ReadOnly MutableClass readOnlyRef = new @ReadOnly MutableClass();

    public void mutableReassign() @Mutable {
        // try to mutate
        field = 3;
        field += 2;
        field++;
        --field;
        this.field++;
        this.field = 3;
        this.field += 2;
        this.field++;

        // reassigning an Assignable Field
        assignable = 3;
        assignable += 2;
        assignable++;
        --assignable;
        this.assignable++;
        this.assignable = 3;
        this.assignable += 2;
        this.assignable++;

        mutableRef = null;
        readOnlyRef = null;
    }

    public void roReassign() @ReadOnly {
        // try to mutate
        field = 3;  // should emit error
        field += 2; // should emit error
        field++;    // should emit error
        --field;    // should emit error
        this.field++;   // should emit error
        this.field = 3; // should emit error
        this.field += 2;    // should emit error
        this.field++;   // should emit error

        // reassigning an Assignable Field
        assignable = 3;
        assignable += 2;
        assignable++;
        --assignable;
        this.assignable++;
        this.assignable = 3;
        this.assignable += 2;
        this.assignable++;

        mutableRef = null;  // should emit error
        readOnlyRef = null; // should emit error
    }

    // Method Calling

    public void roMethodInvoke() @ReadOnly {
        mutableReassign();  // should emit error
        this.mutableReassign(); // should emit error

        this.roReassign();
        roReassign();
    }

    public void mutableRecieverInvoke() @Mutable {
        mutableReassign();
        this.mutableReassign();

        this.roReassign();
        roReassign();
    }

    public void mutateOther() @ReadOnly {
        @Mutable MutableClass instance = new @Mutable MutableClass();
        instance.mutableRecieverInvoke();
        instance.mutableReassign();
        instance.field = 5;
        instance.assignable = 3;
        instance.readOnlyRef.mutableReassign(); // should emit error
        instance.readOnlyRef.assignable = 2;
    }

    public void mutateOetherRO() @Mutable {
        @ReadOnly MutableClass instance = new @ReadOnly MutableClass();
        instance.mutableRecieverInvoke(); //should emit error
        instance.mutableReassign(); //should emit error
        instance.field = 5; //should emit error
        instance.assignable = 3;
        instance.readOnlyRef.mutableReassign(); // should emit error
        instance.mutableRef.field++;
        instance.readOnlyRef.field++;   // should emit error
    }
}
