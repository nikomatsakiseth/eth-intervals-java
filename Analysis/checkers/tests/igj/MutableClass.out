:54: (assignability.invalid)
:55: (assignability.invalid)
:56: (assignability.invalid)
:57: (assignability.invalid)
:58: (assignability.invalid)
:59: (assignability.invalid)
:60: (assignability.invalid)
:61: (assignability.invalid)
:73: (assignability.invalid)
:74: (assignability.invalid)
:80: (method.invocation.invalid)
:81: (method.invocation.invalid)
:101: (method.invocation.invalid)
:107: (method.invocation.invalid)
:108: (method.invocation.invalid)
:109: (assignability.invalid)
:111: (method.invocation.invalid)
:113: (assignability.invalid)
